File: task-10/cat-dog-kill/backend/src/models/Achievement.ts
import mongoose, { Document, Schema } from 'mongoose';

/**
 * 成就模型
 * 定义游戏中所有可解锁的成就
 */
export interface IAchievement extends Document {
  // 成就 ID (唯一标识)
  achievementId: string;
  
  // 成就名称
  name: string;
  
  // 成就描述
  description: string;
  
  // 成就图标 (emoji 或图片 URL)
  icon: string;
  
  // 成就类别
  category: AchievementCategory;
  
  // 成就难度
  difficulty: AchievementDifficulty;
  
  // 解锁条件类型
  conditionType: ConditionType;
  
  // 解锁条件目标值
  conditionTarget: number;
  
  // 成就点数
  points: number;
  
  // 是否隐藏成就
  isHidden: boolean;
  
  // 创建时间
  createdAt: Date;
}

// 成就类别
export enum AchievementCategory {
  // 游戏胜利相关
  VICTORY = 'victory',
  // 任务完成相关
  TASKS = 'tasks',
  // 社交/投票相关
  SOCIAL = 'social',
  // 特殊角色相关
  ROLE = 'role',
  // 连败/连胜相关
  STREAK = 'streak',
  // 收集类
  COLLECTION = 'collection',
  // 特殊成就
  SPECIAL = 'special'
}

// 成就难度
export enum AchievementDifficulty {
  EASY = 'easy',       // 简单 - 1 点
  MEDIUM = 'medium',   // 中等 - 3 点
  HARD = 'hard',       // 困难 - 5 点
  LEGENDARY = 'legendary' // 传奇 - 10 点
}

// 条件类型
export enum ConditionType {
  // 胜利次数
  WINS = 'wins',
  // 失败次数
  LOSSES = 'losses',
  // 游戏场次
  GAMES_PLAYED = 'games_played',
  // 完成任务数
  TASKS_COMPLETED = 'tasks_completed',
  // 投票正确次数
  CORRECT_VOTES = 'correct_votes',
  // 作为狗狗获胜次数
  DOG_WINS = 'dog_wins',
  // 作为猫咪获胜次数
  CAT_WINS = 'cat_wins',
  // 作为狐狸获胜次数
  FOX_WINS = 'fox_wins',
  // 连胜场次
  WIN_STREAK = 'win_streak',
  // 连败场次
  LOSE_STREAK = 'lose_streak',
  // 紧急会议发起次数
  MEETINGS_CALLED = 'meetings_called',
  // 破坏成功次数
  SABOTAGES = 'sabotages',
  // 调查成功次数 (侦探)
  INVESTIGATIONS = 'investigations',
  // 猎人击杀次数
  HUNTER_KILLS = 'hunter_kills',
  // 存活到最后 (狐狸)
  SURVIVALS = 'survivals',
  // 完美任务 (无错误)
  PERFECT_TASKS = 'perfect_tasks',
  // 快速胜利
  SPEED_VICTORY = 'speed_victory',
  // 绝地翻盘
  COMEBACK = 'comeback'
}

const AchievementSchema: Schema = new Schema({
  achievementId: {
    type: String,
    required: true,
    unique: true
  },
  name: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  icon: {
    type: String,
    required: true,
    default: '🏆'
  },
  category: {
    type: String,
    required: true,
    enum: Object.values(AchievementCategory)
  },
  difficulty: {
    type: String,
    required: true,
    enum: Object.values(AchievementDifficulty)
  },
  conditionType: {
    type: String,
    required: true,
    enum: Object.values(ConditionType)
  },
  conditionTarget: {
    type: Number,
    required: true
  },
  points: {
    type: Number,
    required: true,
    default: 1
  },
  isHidden: {
    type: Boolean,
    default: false
  }
}, {
  timestamps: true
});

export default mongoose.model<IAchievement>('Achievement', AchievementSchema);

File: task-10/cat-dog-kill/ios/CatDogKill/Models/Achievement.swift
//
//  Achievement.swift
//  CatDogKill
//
// 成就系统模型
//

import Foundation

// 成就类别
enum AchievementCategory: String, Codable, CaseIterable {
    case victory = "victory"      // 胜利
    case tasks = "tasks"          // 任务
    case social = "social"        // 社交
    case role = "role"            // 角色
    case streak = "streak"        // 连胜
    case collection = "collection" // 收集
    case special = "special"      // 特殊
    
    var localizedName: String {
        switch self {
        case .victory: return "胜利"
        case .tasks: return "任务"
        case .social: return "社交"
        case .role: return "角色"
        case .streak: return "连胜"
        case .collection: return "收集"
        case .special: return "特殊"
        }
    }
    
    var icon: String {
        switch self {
        case .victory: return "🏆"
        case .tasks: return "📋"
        case .social: return "💬"
        case .role: return "🎭"
        case .streak: return "🔥"
        case .collection: return "📦"
        case .special: return "⭐"
        }
    }
}

// 成就难度
enum AchievementDifficulty: String, Codable {
    case easy = "easy"           // 简单
    case medium = "medium"       // 中等
    case hard = "hard"           // 困难
    case legendary = "legendary" // 传奇
    
    var localizedName: String {
        switch self {
        case .easy: return "简单"
        case .medium: return "中等"
        case .hard: return "困难"
        case .legendary: return "传奇"
        }
    }
    
    var color: String {
        switch self {
        case .easy: return "34C759"    // 绿色
        case .medium: return "007AFF"  // 蓝色
        case .hard: return "FF9500"    // 橙色
        case .legendary: return "AF52DE" // 紫色
        }
    }
}

// 成就条件类型
enum ConditionType: String, Codable {
    case wins = "wins"
    case losses = "losses"
    case gamesPlayed = "games_played"
    case tasksCompleted = "tasks_completed"
    case correctVotes = "correct_votes"
    case dogWins = "dog_wins"
    case catWins = "cat_wins"
    case foxWins = "fox_wins"
    case winStreak = "win_streak"
    case loseStreak = "lose_streak"
    case meetingsCalled = "meetings_called"
    case sabotages = "sabotages"
    case investigations = "investigations"
    case hunterKills = "hunter_kills"
    case survivals = "survivals"
    case perfectTasks = "perfect_tasks"
    case speedVictory = "speed_victory"
    case comeback = "comeback"
}

// 成就定义
struct Achievement: Codable, Identifiable {
    let achievementId: String
    let name: String
    let description: String
    let icon: String
    let category: AchievementCategory
    let difficulty: AchievementDifficulty
    let conditionType: ConditionType
    let conditionTarget: Int
    let points: Int
    let isHidden: Bool
    
    var id: String { achievementId }
    
    // 难度颜色
    var difficultyColor: String {
        difficulty.color
    }
}

// 成就进度
struct AchievementProgress: Codable, Identifiable {
    let achievementId: String
    let unlocked: Bool
    let progress: Int
    let unlockedAt: Date?
    
    var id: String { achievementId }
}

// 成就类别信息
struct AchievementCategoryInfo: Codable, Identifiable {
    let id: String
    let label: String
    let count: Int
    
    var category: AchievementCategory? {
        AchievementCategory(rawValue: id)
    }
}

// 用户成就数据
struct UserAchievements: Codable {
    let achievements: [Achievement]
    let progress: [AchievementProgress]
    let totalPoints: Int
    let unlockedCount: Int
    let totalCount: Int
    let completionRate: Int
    
    // 计算某个成就的进度
    func getProgress(for achievementId: String) -> AchievementProgress? {
        progress.first { $0.achievementId == achievementId }
    }
    
    // 是否已解锁某个成就
    func isUnlocked(_ achievementId: String) -> Bool {
        progress.first { $0.achievementId == achievementId }?.unlocked ?? false
    }
    
    // 按类别分组成就
    func groupedByCategory() -> [AchievementCategory: [Achievement]] {
        var grouped: [AchievementCategory: [Achievement]] = [:]
        for achievement in achievements {
            if grouped[achievement.category] == nil {
                grouped[achievement.category] = []
            }
            grouped[achievement.category]?.append(achievement)
        }
        return grouped
    }
}
